direct_consumer: move offset update logic to fetch_next

joe-redpanda opened the pull request:
A fetch is potentially stale until the point in time at which it is returned to the consumer in fetch_next.

This pr moves the update for offsets to only after the final subscription epoch filter has been applied to guarantee correctness of the provided offsets.


backport: this will get backported but we're going to bake the change before putting it through
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [x] none - papercut/not impactful enough to backport
- [ ] v25.3.x
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes
### Bug Fixes
* firm up start offset update logic


<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/75448-->
<!--1dfbb006-1419-488c-9be0-6dba4435015e-->
<!--64fdd4cb-4542-4c36-a00b-c253fdabbf74-->
<!--aedaf5aa-f178-4667-b8c7-f9a6e339f8a0-->
<!--d1aedb96-bcaf-4fe8-90e1-15b3429b3438-->
<!--0b3d2076-276c-4539-a64d-da570782b11b-->
<!--8dd07336-5e5f-4787-bdbd-f58c35ce87e6-->
<!--4b9248e1-c2b4-4c45-837f-51c47881a713-->
<!--37612b93-8de7-4cb2-a997-f8faa2081ea5-->
<!--ceec1e0c-1e0e-4f70-9b28-c307d1ac8d26-->
<!--b904430a-7b70-4199-9fa5-f4cbab1ca853-->
<!--e8c19354-6f50-4f82-92f1-3c4f7c3ac90c-->
<!--eeffc0ab-cb6e-48dc-ad96-7c5d81cd8e4b-->
<!--ba48f20c-5a67-4d79-9363-44fe55b46bfa-->
<!--d4885c30-5651-4bf2-b0c1-72ffeb640e86-->
<!--945d091c-8081-4c07-b509-0ec58b0078d0-->
<!--29e8a18e-bf03-4831-8572-b94a52235beb-->
<!--ab10d4f6-3af9-4170-bf27-578819c674a7-->
<!--82e3a5c7-881e-420d-b7df-0a0b2f7a4e03-->
<!--title-->
## Retry command for [Build#75448](https://buildkite.com/redpanda/redpanda/builds/75448)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_replication_basic@{"shuffle_leadership":true,"source_cluster_spec":{"cluster_type":"redpanda"}}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingMetricsTests.test_link_metrics
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_auto_prefix_trimming@{"source_cluster_spec":{"cluster_type":"redpanda"},"with_failures":false}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_replication_basic@{"shuffle_leadership":false,"source_cluster_spec":{"cluster_type":"redpanda"}}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_auto_prefix_trimming@{"source_cluster_spec":{"cluster_type":"redpanda"},"with_failures":true}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_replication_basic@{"shuffle_leadership":false,"source_cluster_spec":{"cluster_type":"kafka","kafka_quorum":"COMBINED_KRAFT","kafka_version":"3.8.0"}}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_replication_basic@{"shuffle_leadership":true,"source_cluster_spec":{"cluster_type":"kafka","kafka_quorum":"COMBINED_KRAFT","kafka_version":"3.8.0"}}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_auto_prefix_trimming@{"source_cluster_spec":{"cluster_type":"kafka","kafka_quorum":"COMBINED_KRAFT","kafka_version":"3.8.0"},"with_failures":false}
tests/rptest/tests/cluster_linking_e2e_test.py::ShadowLinkingReplicationTests.test_replication_with_failures
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--cd386913-e3bf-477f-b9f3-cdfc4262830a-->
<!--TEST RESULTS-->
<!--2c39ccd8-a32a-48ca-bdc1-e90d212ce45a-->
<!--TEST RESULTS-->
<!--91b855ad-3375-44dc-a560-8ec5fd5336e9-->
<!--TEST RESULTS-->
<!--fd03f754-90d4-431f-bba5-04906be6b6dc-->
<!--TEST RESULTS-->
<!--34045cb1-45d3-48ae-9f39-061a80d0b61a-->
<!--TEST RESULTS-->
<!--1eab620f-0987-465b-ac55-37eb76e04c29-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75448">build#75448</a></summary>


| test_class                           | test_method                                  | test_arguments                                                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                  |
|:-------------------------------------|:---------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync                   | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}, "with_failures": false}      | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c85-5a31-46ff-8ee9-d7b800bcbcbd | FLAKY         | 13/21    | upstream reliability is '93.79432624113475'. current run reliability is '61.904761904761905'. drift is 31.88956 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync |
| ShadowLinkingMetricsTests            | test_link_metrics                            | null                                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c85-5a2e-4740-9258-7a46eb7302d6 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingMetricsTests&test_method=test_link_metrics                     |
| ShadowLinkingMetricsTests            | test_link_metrics                            | null                                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c86-0b6a-4681-9089-024e1e6a3d76 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingMetricsTests&test_method=test_link_metrics                     |
| ShadowLinkingReplicationTests        | test_auto_prefix_trimming                    | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}, "with_failures": false}      | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c85-5a2f-4b0e-9953-7435dad72707 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_auto_prefix_trimming         |
| ShadowLinkingReplicationTests        | test_auto_prefix_trimming                    | {"source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}, "with_failures": false}      | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c86-0b6c-4874-bb34-2ef562459b73 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_auto_prefix_trimming         |
| ShadowLinkingReplicationTests        | test_auto_prefix_trimming                    | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": false}                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c85-5a30-4271-aaf4-fb3bf374bb54 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_auto_prefix_trimming         |
| ShadowLinkingReplicationTests        | test_auto_prefix_trimming                    | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": false}                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c86-0b6d-4647-8590-761d2bdd10ca | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_auto_prefix_trimming         |
| ShadowLinkingReplicationTests        | test_auto_prefix_trimming                    | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true}                                                                | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c85-5a33-4dcd-872c-6097fa4cc086 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_auto_prefix_trimming         |
| ShadowLinkingReplicationTests        | test_auto_prefix_trimming                    | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true}                                                                | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c86-0b70-4169-bd9a-d7850ce5f958 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_auto_prefix_trimming         |
| ShadowLinkingReplicationTests        | test_replication_basic                       | {"shuffle_leadership": false, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c85-5a34-40cf-ae3b-e35bf2f2eb94 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic            |
| ShadowLinkingReplicationTests        | test_replication_basic                       | {"shuffle_leadership": false, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c86-0b71-4e92-b1f2-f68e570beceb | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic            |
| ShadowLinkingReplicationTests        | test_replication_basic                       | {"shuffle_leadership": false, "source_cluster_spec": {"cluster_type": "redpanda"}}                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c85-5a35-4dd6-a24b-ff00d2b30e89 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic            |
| ShadowLinkingReplicationTests        | test_replication_basic                       | {"shuffle_leadership": false, "source_cluster_spec": {"cluster_type": "redpanda"}}                                                          | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c86-0b72-4e64-a66f-9ae3497c14bd | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic            |
| ShadowLinkingReplicationTests        | test_replication_basic                       | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}}  | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c85-5a37-42cc-b4c7-fc2ca44a65d0 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic            |
| ShadowLinkingReplicationTests        | test_replication_basic                       | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}}  | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c86-0b74-4602-9fc9-5a948bfc6e7f | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic            |
| ShadowLinkingReplicationTests        | test_replication_basic                       | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}}                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c85-5a38-44b4-b92e-e524db91e0f7 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic            |
| ShadowLinkingReplicationTests        | test_replication_basic                       | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}}                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c86-0b67-45a8-9d63-998b5152b9e9 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic            |
| ShadowLinkingReplicationTests        | test_replication_timestamps_match            | {"source_cluster_spec": {"cluster_type": "redpanda"}, "timestamp_type": "CreateTime"}                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c85-5a2e-4740-9258-7a46eb7302d6 | FLAKY         | 20/21    | upstream reliability is '95.14563106796116'. current run reliability is '95.23809523809523'. drift is -0.09246 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_timestamps_match |
| ShadowLinkingReplicationTests        | test_replication_with_failures               | null                                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c85-5a31-46ff-8ee9-d7b800bcbcbd | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures    |
| ShadowLinkingReplicationTests        | test_replication_with_failures               | null                                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c86-0b6f-4d4f-b7f1-6d4eaf25bb36 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                                   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_with_failures    |
| SegmentMsTest                        | test_segment_rolling_with_retention_consumer | null                                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c85-5a35-4dd6-a24b-ff00d2b30e89 | FLAKY         | 18/21    | upstream reliability is '94.56118665018542'. current run reliability is '85.71428571428571'. drift is 8.8469 and the allowed drift is set to 50. The test should PASS    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SegmentMsTest&test_method=test_segment_rolling_with_retention_consumer      |
| PartitionMovementTest                | test_static                                  | {"num_to_upgrade": 0}                                                                                                                       | integration | https://buildkite.com/redpanda/redpanda/builds/75448#019a3c86-0b68-4129-ab3c-0a4318b48637 | FLAKY         | 16/21    | upstream reliability is '87.01923076923077'. current run reliability is '76.19047619047619'. drift is 10.82875 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionMovementTest&test_method=test_static                               |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/75826">build#75826</a></summary>


| test_class                    | test_method                          | test_arguments                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                             |
|:------------------------------|:-------------------------------------|:----------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ReplicatedMetastoreTest       | TestBasicRemoveTopics                |                                                                                   | unit        | https://buildkite.com/redpanda/redpanda/builds/75826#019a5eca-8850-4ab4-91ab-c03c7d8ca487 | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                          |
| ShadowLinkingReplicationTests | test_replication_basic               | {"shuffle_leadership": true, "source_cluster_spec": {"cluster_type": "redpanda"}} | integration | https://buildkite.com/redpanda/redpanda/builds/75826#019a5f36-98e8-47c5-9b8b-0749ca2788df | FLAKY         | 19/21    | upstream reliability is '98.19148936170212'. current run reliability is '90.47619047619048'. drift is 7.7153 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic       |
| DataMigrationsApiTest         | test_creating_and_listing_migrations | null                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/75826#019a5f36-98ef-489b-913d-e1d49d14869d | FLAKY         | 19/21    | upstream reliability is '98.13620071684588'. current run reliability is '90.47619047619048'. drift is 7.66001 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DataMigrationsApiTest&test_method=test_creating_and_listing_migrations |
| MountUnmountIcebergTest       | test_simple_remount                  | {"cloud_storage_type": 1}                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/75826#019a5f56-8c48-4a17-8e35-0d209699350c | FLAKY         | 17/21    | upstream reliability is '91.97860962566845'. current run reliability is '80.95238095238095'. drift is 11.02623 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount                |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77609">build#77609</a></summary>


| test_class                         | test_method                   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                   |
|:-----------------------------------|:------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ReplicatedMetastoreTest            | TestBasicRemoveTopics         |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/77609#019b0507-1d4f-4e19-b134-ab607e210ee1 | FAIL          | 0/1      |                                                                                                                                                                                                                                                                                                                                                                                                     |                                                                                                                                                                                |
| ControllerLogLimitMirrorMakerTests | test_mirror_maker_with_limits | null             | integration | https://buildkite.com/redpanda/redpanda/builds/77609#019b0533-0a85-47bc-bb39-9e9da6e76c03 | FLAKY         | 20/21    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0177, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.0521, p1=0.3432, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ControllerLogLimitMirrorMakerTests&test_method=test_mirror_maker_with_limits |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77625">build#77625</a></summary>


| test_class                 | test_method                              | test_arguments                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                      |
|:---------------------------|:-----------------------------------------|:-----------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodesDecommissioningTest   | test_decommissioning_rebalancing_node    | {"shutdown_decommissioned": false} | integration | https://buildkite.com/redpanda/redpanda/builds/77625#019b05da-30ef-40ac-9406-20a7ededdbc2 | FLAKY         | 9/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.1046, p0=0.6688, reject_threshold=0.0100. adj_baseline=0.2821, p1=0.1792, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_rebalancing_node      |
| NodesDecommissioningTest   | test_decommissioning_rebalancing_node    | {"shutdown_decommissioned": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/77625#019b05da-30f0-4374-85d5-c492f17ec5f3 | FLAKY         | 9/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.1046, p0=0.6688, reject_threshold=0.0100. adj_baseline=0.2821, p1=0.1792, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_rebalancing_node      |
| TxAtomicProduceConsumeTest | test_basic_tx_consumer_transform_produce | {"with_failures": true}            | integration | https://buildkite.com/redpanda/redpanda/builds/77625#019b05da-5cc1-4f97-a8eb-1ff7a1e54cd4 | FLAKY         | 30/31    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0088, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.0262, p1=0.4516, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=TxAtomicProduceConsumeTest&test_method=test_basic_tx_consumer_transform_produce |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/77668">build#77668</a></summary>


| test_class               | test_method                           | test_arguments                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                                 |
|:-------------------------|:--------------------------------------|:----------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| NodesDecommissioningTest | test_decommissioning_rebalancing_node | {"shutdown_decommissioned": true} | integration | https://buildkite.com/redpanda/redpanda/builds/77668#019b09b0-a7fd-4a94-98f9-6e28a3e61dc6 | FLAKY         | 8/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.1030, p0=0.2757, reject_threshold=0.0100. adj_baseline=0.2784, p1=0.4425, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodesDecommissioningTest&test_method=test_decommissioning_rebalancing_node |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/78229">build#78229</a></summary>


| test_class                          | test_method         | test_arguments              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                                                                                                                                                                                                                                              | test_history                                                                                                                                                     |
|:------------------------------------|:--------------------|:----------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MountUnmountIcebergTest             | test_simple_remount | {"cloud_storage_type": 1}   | integration | https://buildkite.com/redpanda/redpanda/builds/78229#019b3832-437a-427a-a961-8d913f2340df | FLAKY         | 8/11     | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.1799, p0=0.5606, reject_threshold=0.0100. adj_baseline=0.4485, p1=0.1012, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=MountUnmountIcebergTest&test_method=test_simple_remount        |
| WriteCachingFailureInjectionE2ETest | test_crash_all      | {"use_transactions": false} | integration | https://buildkite.com/redpanda/redpanda/builds/78229#019b3836-152d-441c-9809-4d71fc4d9c89 | FLAKY         | 10/11    | Test PASSES after retries.[No significant increase in flaky rate](https://redpandadata.atlassian.net/wiki/spaces/DEVPROD/pages/1419247661/Statistical+retry+failure+reasons#%5BinlineExtension%5DCase-C%3A-Test-passes-with-a-rate-not-significantly-worse-than-baseline-(run-passes))(baseline=0.0709, p0=1.0000, reject_threshold=0.0100. adj_baseline=0.1979, p1=0.1103, trust_threshold=0.5000) | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all |
</details>
<!--/body-->

joe-redpanda commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -103,6 +104,59 @@ void direct_consumer::filter_stale_subscriptions(
+              "invoke only after filtering, all subscriptions should be "
+              "consistent");
+            auto& subscription = maybe_subscription->get();
+            // subscription.last_known_source_offsets =
```
oops

bharathv commented:
The failures here seem related to the change?

joe-redpanda commented:
/ci-repeat 1
tests/rptest/tests/cluster_linking_e2e_test.py

joe-redpanda committed:
dc/tests: change asserts to anticipate offset only

direct_consumer_fixture_test now only asserts on initial conditions in
one test to reduce redundancy. The initial assert is changed to permit
offset only updates, but restrict those updates to at max one per
partition.

direct_consumer_test is changed to no longer assert on empty fetches.

joe-redpanda committed:
direct_consumer: permit offset only fetches

This commit does four things
1. fetch data will now be added to the queue even if it has no batches
2. offsets will be updated at the point in time at which fetch_next is
   called
3. a new filter is applied to remove fetches which contain no new
   information (all offsets are the same)
4. fetch next will retry fetching from the data_queue if the filters
   have removed everything from the resultant fetch

joe-redpanda commented on src/v/kafka/client/direct_consumer/api_types.h:
```diff
@@ -68,7 +68,7 @@ struct fetched_partition_data {
     std::optional<chunked_vector<aborted_transaction>> aborted_transactions;
     subscription_epoch subscription_epoch;
-    size_t size_bytes;
+    size_t size_bytes{0};
```
this will get fixed in a rebase against dev

joe-redpanda committed:
direct_consumer: clarity fixes

The results of partitioning were incorrectly named in
filter_stale_subscriptions.

Fix the names and additionally pull the iterators from the subspan
rather than doing wasteful (and dangerous) iterator math.

joe-redpanda commented:
/ci-repeat 1
tests/rptest/tests/cluster_linking_e2e_test.py
tests/rptest/direct_consumer_tests/direct_consumer_test.py

joe-redpanda commented:
/ci-repeat 1
tests/rptest/tests/cluster_linking_e2e_test.py
tests/rptest/direct_consumer_tests/direct_consumer_test.py

joe-redpanda commented:
/ci-repeat 1
tests/rptest/tests/cluster_linking_e2e_test.py
tests/rptest/direct_consumer_tests/direct_consumer_test.py

joe-redpanda commented:
/ci-repeat 1
tests/rptest/tests/cluster_linking_e2e_test.py
tests/rptest/direct_consumer_tests/direct_consumer_test.py

joe-redpanda committed:
direct_consumer/fetcher: extract process logic

Fetcher is remarkably error prone to work on.
To mitigate this, this commit splits the logic for processing fetch
responses into new delegated functions.

1. do_process_partition_response: a sync static method which is
   responsible for taking a given fetch and determining what should be
   done with it
   - retriable errors -> update metadata
   - out of bounds -> reset offsets
   - unknown error -> bubble to caller
   - data fetch -> return data
   - offset only fetch -> return offsets
2. process_partition_response: an async wrapper for
   do_process_partition_response which updates the fetcher local state
   and incorporates the results into the resultant fetch response

joe-redpanda committed:
direct_consumer/fetcher: add fetcher test

Adds fetcher unit tests to ensure the decision logic in
do_process_partition_response is per expectations.

joe-redpanda committed:
direct_consumer/api_types.h

Add reasonable initializers to fetched_partition_data. This is not a
required correctness change, instead this is meant to preempt a writer
from forgetting to set a value when filling in the fields of
fetched_partition_data.

Adds are_offfsets_equal to source_partition_offsets which will check if
tracked offsets are the same or different over time.

joe-redpanda committed:
direct_consumer: utility find subscription funcs

Adds utility functions to find subscriptions, returning the reference as
an optional on reference wrapper. Updates usages of these to use the
helpers for cleanliness

joe-redpanda commented:
/ci-repeat 1
tests/rptest/tests/cluster_linking_e2e_test.py
tests/rptest/direct_consumer_tests/direct_consumer_test.py

michael-redpanda commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -718,6 +598,207 @@ fetcher::process_fetch_response(
+        vassert(
+          response_records.size() != 0,
+          "nonzero size should have nonzero records");
+    }
```
Is this risky at all?  Could a misbehaving broker cause us to crash in this situation?

michael-redpanda commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -103,6 +130,99 @@ void direct_consumer::filter_stale_subscriptions(
+                      "current: {}, found: {}",
+                      subscription.last_known_source_offsets.last_stable_offset,
+                      spo.last_stable_offset);
+                }
```
I'd argue all of these could be warn.

bharathv commented:
/dt

bharathv commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -551,61 +563,88 @@ fetcher::process_fetch_response(
+                    // these still go on the queue as they probably contain
+                    // information about a prefix truncation
+                    vlog(
+                      logger().info,
```
is this worth info level? doesn't seem like it.

bharathv commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -551,61 +563,88 @@ fetcher::process_fetch_response(
+                      topic_data.topic,
+                      part_data.partition_id);
+                    // maybe_update_fetch_offset(topic_data.topic,
+                    // part_data.partition_id, )
```
remove?

joe-redpanda replied:
oops, thank you

bharathv commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -551,61 +563,88 @@ fetcher::process_fetch_response(
+            vassert(
+              topic_data.total_bytes == 0,
+              "bad size accounting, if fetched_topic_data has no partitions, "
+              "its size must necessarily be zero");
```
consider adding a bit more context, otherwise hard to pin point exactly what is happening.

bharathv commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -103,6 +130,99 @@ void direct_consumer::filter_stale_subscriptions(
+                  topic_data.topic, partition_data.partition_id);
+                vassert(
+                  maybe_subscription.has_value(),
+                  "invoke only after filtering, all subscriptions should be "
```
context

bharathv commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -37,29 +40,53 @@ direct_consumer::direct_consumer(
+            filter_stale_subscriptions(response_to_filter);
+
+            // update offset info and filter old fetches
+            update_offsets_filter_redundant_fetches(response_to_filter);
```
wondering if we can do a one pass filter both (club into a single lambda that evaluates both?)

joe-redpanda replied:
Definitely could, I did this in two phases for readability. Happy either way though, I could just subdivide the lambda into a a few helpers to logically subdivide

joe-redpanda replied:
added

bharathv commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -103,6 +130,99 @@ void direct_consumer::filter_stale_subscriptions(
+                    vlog(
+                      _cluster->logger().error,
+                      "last known stable should never move backward, "
+                      "current: {}, found: {}",
```
topic partition info?

bharathv reviewed (commented):
lgtm, mostly minor comments, thanks for the cleanup

> This pr moves the update for offsets to only after the final subscription epoch filter has been applied to guarantee correctness of the provided offsets.

can you detail the conditions under which correctness is compromised (for posterity), AFAICT it is very very rare

dotnwat commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -37,34 +40,59 @@ direct_consumer::direct_consumer(
 
 ss::future<fetches>
 direct_consumer::fetch_next(std::chrono::milliseconds timeout) {
+    static constexpr auto jitter = std::chrono::milliseconds{1};
```
we have a jitter utility that you can use. also pretty sure lowres clock has a resolution less than 1 millisecond...

joe-redpanda replied:
ack on the utility

Will probably swap to a simple std::max(deadline, min_timeout) because the original goal was to give the call to pop some reasonable minimum timeout.

per the above: lowres clock has a granularity of around task_quota, and task quota is .5ms. I'm thinking that 2 task quotas is still a reasonable minimum timeout

dotnwat commented on src/v/kafka/client/direct_consumer/api_types.h:
```diff
@@ -59,14 +59,15 @@ struct topic_assignment {
-    kafka::leader_epoch leader_epoch;
-    kafka::offset start_offset;
-    kafka::offset high_watermark;
-    kafka::offset last_stable_offset;
-    chunked_vector<model::record_batch> data;
+    model::partition_id partition_id{-1};
+    kafka::leader_epoch leader_epoch{-1};
+    kafka::offset start_offset{-1};
+    kafka::offset high_watermark{-1};
+    kafka::offset last_stable_offset{-1};
+    chunked_vector<model::record_batch> data{};
     kafka::error_code error = kafka::error_code::none;
-    std::optional<chunked_vector<aborted_transaction>> aborted_transactions;
+    std::optional<chunked_vector<aborted_transaction>> aborted_transactions{
+      std::nullopt};
```
> Add initializers to fetched_partition_data.

weren't these already all initialized before this change? if so, i think the commit needs to describe why either a different initial value is needed or why the original initializaiton wasn't sufficient.

joe-redpanda replied:
added some description to both the commit and code.

the goal was to prevent someone from default initializing the struct and having it preloaded valid offsets, partition id, etc. The concern being that partition_id{-1} is definitely wrong, where a default of partiton_id{0} could be subtly wrong

dotnwat replied:
makes sense, and +1 to defensive programming! but definitely add a comment. if i saw this code without a comment, i'd be tempted to just remove all of it because the default initializaiton values for the named types are all negative values and the other two have default initialization values that are no different than what is in this commit.

dotnwat commented on src/v/kafka/client/direct_consumer/api_types.h:
```diff
@@ -90,6 +91,13 @@ struct source_partition_offsets {
     // The timestamp that the fetch response was received by the client
     ss::lowres_clock::time_point last_offset_update_timestamp{};
+
+    // check all except timestamp
```
comment should say why we aren't checking the timestamp.

joe-redpanda replied:
Yeah on reconsideration I'm not sure why I wrote an == that ignores a field. I'm going to swap to an 'are_offsets_equal" method for clarity

dotnwat replied:
> Yeah on reconsideration I'm not sure why I wrote an == that ignores a field. I'm going to swap to an 'are_offsets_equal" method for clarity

well i think the next question is going to be why are we comparing timestamp equality? is it for tests? otherwise it seems niche to be comparing timestamps given the nature of time.

joe-redpanda replied:
Removed equality operator in exchange for a method to compare offsets for clarity sake. I feel like the equality operator could go either way on whether or not to check timestamp, but a custom method is self documenting.

dotnwat commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -718,6 +598,207 @@ fetcher::process_fetch_response(
+  chunked_vector<model::record_batch> response_batches,
+  size_t response_size,
+  fetcher::epoch_set epoch_set) {
+    // this was done to keep this method synchronous
```
can you explain what this vassert has to do with the method being synchronous?

joe-redpanda replied:
```
    // pull the records from the response if present
    chunked_vector<model::record_batch> response_records{};
    size_t response_size{0};
    if (
      partition_response.records.has_value()
      && !partition_response.records->is_end_of_stream()) {
        response_size = partition_response.records->size_bytes();
        response_records = co_await reader_to_chunked_vector(
          std::move(partition_response.records.value()));
    }
```

the deserialization of records is an async operation which means if I want do_process_partition_response to be synchronous, the caller will have to perform the deserialization. This vassert is a quick check that the caller isn't asking for a decision to be made on a response which has not had its records extracted.

dotnwat replied:
i think that deserves a comment. it makes sense, but it's also non-obvious

joe-redpanda replied:
added in latest

dotnwat commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -718,6 +598,207 @@ fetcher::process_fetch_response(
+    vassert(
+      maybe_epoch_set.has_value(),
+      "tp should be found in snapshotted epochs if the response is "
+      "epoch consistent");
```
if the assertion fires, would you expect that someone would be able to debug with only this static string?

joe-redpanda replied:
updated, with what is hopefully better context

Detailed context:
The original fetcher code was a lot of lines like

```cpp
auto topic_iterator = topic.map.find(topic);
if (topic_iterator == topic_map.end()) {
return;
}
auto partition_iterator = partition_map.find(partition);
if(partition_iterator == partition_map.end()) {
return;
}
auto& some_value = *partition_iterator;
```

I added a consistency filter which essentially does on response processing:
for a given partition response to not be filtered out:
1. the partition must have been assigned when the fetch began
2. the partition must be assigned when the response is processed
3. the assignment epoch of the above two must be the same

By checking this all once and at the beginning of fetch processing (and holding a lock on the assignments map) we can skip many of the validity checks.

The above is an example, if the epoch set isn't found, we violate #1 of our consistency check

These are still guarded in vasserts though

dotnwat replied:
👍

dotnwat commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -718,6 +598,207 @@ fetcher::process_fetch_response(
+        vassert(
+          response_records.size() != 0,
+          "nonzero size should have nonzero records");
+    }
```
isn't it only to our advantage that we record more context in these never-should-happen log messages? like we get an assertion that fires what will `zero size should have zero records` tell us when debugging? there is a co-await above this too, so the log message at the beginning of the function that provides some context may not be able to be correlated with the assertion?

joe-redpanda replied:
ack and understood. These I removed because I think they were a bit too paranoid

dotnwat reviewed (commented):
my only concern is that this seems like a non-trivial change that is going to be backported?

joe-redpanda commented:
> my only concern is that this seems like a non-trivial change that is going to be backported?

Agreed. My thought was to leave this to bake for some time before performing the back-port.

joe-redpanda commented:
> lgtm, mostly minor comments, thanks for the cleanup
> 
> > This pr moves the update for offsets to only after the final subscription epoch filter has been applied to guarantee correctness of the provided offsets.
> 
> can you detail the conditions under which correctness is compromised (for posterity), AFAICT it is very very rare

Short:
a lot of cached fetches + a prefix truncation.

Your start offset change will be visible right away, where all other updates will be visible whenever the queue gets burned down.

Long:
Start offsets gets updated at the time a fetch is performed instead of when it passes by direct consumer, while all other data is made visible at the time of direct_consumer::fetch_next

so for tp topic/1 you can have
fetch -> start: 0, hwm: 100
fetch -> start: 0, hwm: 200
fetch -> start: 200, hwm: 300

direct_consumer::get_start_offset("topic/1") -> 200
direct_consumer::fetch_next -> start: 0, hwm: 100

25.3 start offset was only getting used for metrics so this wasn't really an important issue. We decided the above was fine for the release but correctness should be fixed s.t. we can depend on the correct ordering of offsets.

joe-redpanda commented:
dev rebase

joe-redpanda commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -37,54 +40,90 @@ direct_consumer::direct_consumer(
+
+              vassert(
+                maybe_subscription,
+                "nullopt implies stale and should have been handled above");
```
will edit this to be independent of above / below etc.

joe-redpanda committed:
direct_consumer: vassert clarity fixes

Clarifies why a vassert is firing in direct_consumer:
Direct consumers filtering is order dependent to reduce the amount of
code spent checking nullopt

joe-redpanda committed:
direct_consumer/fetcher: vassert clarity fixes

Adds clarity to the meaning and implication of consistent partitions in
vasserts.

Namely, the code is written to check whether a partition is consistent
before operating on it, allowing us to skip most checks on iterators
to the end of a collection and nullopts from helper getting methods.

This was done to significantly cut down on invalid entry checks.

joe-redpanda commented on src/v/kafka/client/direct_consumer/fetcher.cc:
```diff
@@ -292,6 +301,9 @@ bool fetcher::maybe_update_fetch_offset(
     }
 
     auto& fetch_state = maybe_fetch_state->get();
+    if (fetch_state.fetch_offset == kafka::next_offset(last_received)) {
```
update to check hwm as well

joe-redpanda replied:
mm no
```
A partition is only included in an incremental FetchRequest if:
    The client wants to notify the broker about a change to the partition's maxBytes, fetchOffset, or logStart
    The partition was not included in the incremental fetch session before, but the client wants to add it.
    The partition is in the incremental fetch session, but the client wants to remove it.
```
this implies that 'should_skip' isn't correct, though, because should_skip checks both incremental_include and whether or not last_fetched is >= hwm.

We are probably over eagerly sending data in our requests for incremental fetch.

I'll cut a ticket to shave this down to only what is defined in the KIP

joe-redpanda replied:
For any future reader:

The kip is saying only include an update to the brokers in a fetch request for edits to the membership of the fetch session OR if the last seen offset has been updated.

We are currently including the partition in the fetch request if we have any delta between the HWM and the last seen.

This probably isn't that big of an issue because a continuously growing partition will probably be included in the fetch request anyway because we should feasibly always receive new data for it if its growing.

When this will be an actual performance hit is when the partition hwm is higher than last seen but for whatever reason batches aren't included in the next fetch. The broker already understands that it owes the consumer data, but the consumer will add that partition to the next request anyway (and needlessly)

bharathv commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -37,54 +40,90 @@ direct_consumer::direct_consumer(
-        if (maybe_response_to_filter.has_error()) {
-            co_return maybe_response_to_filter;
-        }
+        // we'll keep attempting to pluck from the queue until the timeout is
+        // exhausted
+        while (ss::lowres_clock::now() < deadline) {
+            // either the remaining timeout or a small but reasonable minimum
+            // timeout
```
may  I ask why this change is needed?  It seems like this timeout should be honored within the data_queue API if nothing is available to be fetched.

joe-redpanda replied:
With filtering its possible that the next batch of fetch data from the queue is empty. The loop will continuously grab new fetches off the queue until it finds one that isn't entirely stale.

The deadline fiddling done here is so we don't join the cv's waiter queue with an already expired / imminently expiring timeout. If we're going through the work of calling into fetch_next, imo we should give the operation at least a task_quota to work with

bharathv commented on src/v/kafka/client/direct_consumer/direct_consumer.cc:
```diff
@@ -122,7 +122,8 @@ void direct_consumer::filter_fetch_data(
               vassert(
                 maybe_subscription,
-                "nullopt implies stale and should have been handled above");
+                "offset filtering requires that unassigned subscriptions have "
+                "already been filtered out");
```
nit: do we need to dump any debug state in the assert output

joe-redpanda replied:
I don't think so. This is a paranoid assert. I can feasibly only see this assert firing if a developer changes direct consumer, and the message is fairly clear in so far as I think it indicates
"you goofed up, this used to filter expired subscriptions and now it doesn't"

bharathv reviewed (commented):
lgtm modulo one question around a loop with timer.

as discussed offline, we should probably skip the backport or have some extra baketime before backporting.

joe-redpanda commented on src/v/kafka/client/direct_consumer/tests/direct_consumer_test.cc:
```diff
@@ -654,9 +638,12 @@ TEST_F(consumer_test_mock, TestOffsetResetPolicy) {
-    // fetch should be empty, because the fetch offset was reset to latest
-    ASSERT_EQ(data.value().size(), 0);
+    for (auto& topic_data : data.value()) {
+        ASSERT_EQ(topic_data.total_bytes, 0)
```
the next line has a message. There cannot be a semicolon between

bharathv reviewed (approved):
lgtm
